cmd/stellar-rpc/internal/ingest/backfill.go

```diff
@@ -0,0 +1,278 @@
+// Used to fill local DB backwards towards older ledgers
+// Returns the rightmost ledger
+func runBackfillBackwards(callerCtx context.Context, logger *supportlog.Entry, ledgerRW db.ReadWriter, dsInfo DatastoreInfo, lBound uint32, rBound uint32) error {
+	for rChunkBound := rBound; rChunkBound >= lBound; {
+		if err := callerCtx.Err(); err != nil {
+			return err
+		}
+		// Create temporary backend for backwards-filling chunks
+		tempBackend, err := makeBackend(dsInfo)
+		if err != nil {
+			return fmt.Errorf("couldn't create backend: %w", err)
+		}
+		defer tempBackend.Close()
+
+		lChunkBound := max(lBound, rChunkBound-ChunkSize+1)
+		logger.Infof("Backwards backfill: backfilling ledgers [%d, %d]", lChunkBound, rChunkBound)
+		backfillRange := ledgerbackend.BoundedRange(lChunkBound, rChunkBound)
+		if err := tempBackend.PrepareRange(callerCtx, backfillRange); err != nil {
+			return fmt.Errorf("couldn't prepare range [%d, %d]: %w", lChunkBound, rChunkBound, err)
+		}
+
+		tx, err := ledgerRW.NewTx(callerCtx)
+		if err != nil {
+			return fmt.Errorf("couldn't create local db write tx: %w", err)
+		}
+		defer tx.Rollback()
+
+		if err := fillChunk(callerCtx, dsInfo, tx, &tempBackend, lChunkBound, rChunkBound); err != nil {
+			return fmt.Errorf("couldn't fill chunk [%d, %d]: %w", lChunkBound, rChunkBound, err)
+		}
+		logger.Infof("Backwards backfill: committed ledgers [%d, %d]", lChunkBound, rChunkBound)
+
+		if lChunkBound == lBound {
+			break
+		}
+		rChunkBound = lChunkBound - 1
+	}
+	return nil
+}
+
+// Backfills the local DB with ledgers in [lBound, rBound] from the cloud datastore
+// Used to fill local DB backwards towards the current ledger tip
+func runBackfillForwards(callerCtx context.Context, logger *supportlog.Entry, ledgerRW db.ReadWriter, dsInfo DatastoreInfo, lBound uint32, rBound uint32) error {
+	for lChunkBound := lBound; lChunkBound <= rBound; lChunkBound += ChunkSize {
+		if err := callerCtx.Err(); err != nil {
+			return err
+		}
+		rChunkBound := min(rBound, lChunkBound+ChunkSize-1)
+		logger.Infof("Forwards backfill: backfilling ledgers [%d, %d]", lChunkBound, rChunkBound)
+		backfillRange := ledgerbackend.BoundedRange(lChunkBound, rChunkBound)
+		if err := dsInfo.backend.PrepareRange(callerCtx, backfillRange); err != nil {
+			return fmt.Errorf("couldn't prepare range [%d, %d]: %w", lChunkBound, rChunkBound, err)
+		}
+
+		tx, err := ledgerRW.NewTx(callerCtx)
+		if err != nil {
+			return fmt.Errorf("couldn't create local db write tx: %w", err)
+		}
+
+		if err := fillChunk(callerCtx, dsInfo, tx, nil, lChunkBound, rChunkBound); err != nil {
+			return fmt.Errorf("couldn't fill chunk [%d, %d]: %w", lChunkBound, rChunkBound, err)
+		}
+		logger.Infof("Forwards backfill: committed ledgers [%d, %d]", lChunkBound, rChunkBound)
+
+	}
+	return nil
+}
+
+func fillChunk(callerCtx context.Context, dsInfo DatastoreInfo, tx db.WriteTx, tempBackend *ledgerbackend.LedgerBackend, left uint32, right uint32) error {
```

Comment:
this should just be an interface

```suggestion
func fillChunk(callerCtx context.Context, dsInfo DatastoreInfo, tx db.WriteTx, tempBackend ledgerbackend.LedgerBackend, left uint32, right uint32) error {
```